### PR TITLE
Empty state button rules

### LINKED ISSUE #1222: [Enhancement] Button rules in MessagingStates / EmptyStates
**Is your enhancement request related to a problem? Please describe.**
At this moment of time the component all combinations of buttons and placements of buttons, like 2 attentionlvl 1 buttons placed as number 2 and 3. This ends up in 

**Describe the solution you'd like**
We would like to have clearly defined rules of what is allowed in a MessagingState / Emptystate:

1. If you have a button that is attentionlvl1, this can only be placed at the first place - and there can only be one!
2. All the other buttons should be attentionlvl3 and the first button can also be attentionlvl3

**Describe alternatives you've considered**
Nothing

**Additional context**
If people have implemented another button like attentionlvl2 or 4 it should default to a attentionlvl3.
Button rules

### PR DESCRIPTION
**Pull request by MadsBuchmann**
This PR closes #1222. 

## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] cookbook application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #1222 

## What is the new behavior?
When a `emptyStateComponent` has `button`-children all of them will have their attention level set to 3; except the first one with attention level 1. This also applies if buttons are dynamically added or removed from the `emptyStateComponent`. 

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

## Other information
This is a draft as I still have to add tests for adding and removing button children from the emptyStateComponent. I however need a bit of help with this; see comments.

### CODE REVIEW

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts [hunk 1]
```diff
@@ -35,4 +40,32 @@ describe('EmptyStateComponent', () => {
...
+      expect(numberLvl1Buttons).toEqual(1);
+      expect(numberLvl3Buttons).toEqual(3);
+    });
+
+    it('should keep the attention level on the first button with attention level 1', () => {
+      expect(buttons.findIndex((button) => button.isAttentionLevel1)).toEqual(0);
+    });
+
+    it('should enforce attention level rules when a button is added', () => {
+      // TODO: implement test...
```
Comment (MadsBuchmann) [inline libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts:64]: I could use some help on how to go about this - I've been searching for other tests that do something similar, but can't seem to find any. Any help would be much appreciated 🥳

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts [hunk 2]
```diff
@@ -35,4 +40,32 @@ describe('EmptyStateComponent', () => {
...
+    });
+
+    it('should enforce attention level rules when a button is removed', () => {
+      // TODO: implement test...
```
Comment (MadsBuchmann) [inline libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts:68]: I could use some help on how to go about this - I've been searching for other tests that do something similar, but can't seem to find any. Any help would be much appreciated 🥳

File: apps/cookbook/src/app/examples/empty-state-example/empty-state-example.component.html [hunk 3]
```diff
@@ -42,3 +42,12 @@
 >
 </kirby-empty-state>
 <hr />
+<kirby-empty-state
+  iconName="help"
+  title="Button attention levels"
+  subtitle="All slotted buttons will have attention level 3. Except the first button with level 1. "
```
Comment (jakobe) [inline apps/cookbook/src/app/examples/empty-state-example/empty-state-example.component.html:48]: Whitespace 😋 
```suggestion
  subtitle="All slotted buttons will have attention level 3. Except the first button with level 1."
```

Comment (jakobe) [inline apps/cookbook/src/app/examples/empty-state-example/empty-state-example.component.html:48]: Consider wording:
```suggestion
  subtitle="All slotted buttons will be rendered with attention level 3, except the first button if it is configured with attention level 1. "
```

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts [hunk 4]
```diff
@@ -22,6 +24,9 @@ describe('EmptyStateComponent', () => {
       subtitle="You don't have any items. Call support to add some items to your account."
     >
       <button kirby-button>Call support</button>
+      <button kirby-button attentionLevel='2'> Mail support</button>
+      <button kirby-button attentionLevel='3'> Get directions </button>
+      <button kirby-button attentionLevel='1'> Mail support</button>
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts:29]: Whitespace 😋 
```suggestion
      <button kirby-button attentionLevel='2'>Mail support</button>
      <button kirby-button attentionLevel='3'>Get directions</button>
      <button kirby-button attentionLevel='1'>Mail support</button>
```

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts [hunk 5]
```diff
@@ -35,4 +40,32 @@ describe('EmptyStateComponent', () => {
...
   });
+
+  describe('Attention level rules for buttons', () => {
+    let buttons: ButtonComponent[];
+
+    beforeEach(() => {
+      buttons = spectator.queryAll(ButtonComponent);
+    });
+
+    it('should change all buttons to attention level 3 except 1', () => {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts:51]: Consider wording:
```suggestion
    it('should change all buttons to attention level 3 except the 1st', () => {
```

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.ts [hunk 6]
```diff
@@ -1,13 +1,58 @@
...
+  @ContentChildren(ButtonComponent)
+  private slottedButtons: QueryList<ButtonComponent>;
+
+  private ngUnsubscribe$ = new Subject<void>();
+
+  ngAfterContentInit() {
+    this.enforceAttentionLevelRules();
+
+    this.slottedButtons.changes
+      .pipe(takeUntil(this.ngUnsubscribe$), delay(0))
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/empty-state/empty-state.component.ts:34]: No need to unsubscribe as this is handled by Angular when the component gets destroyed
```suggestion
```

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.ts [hunk 7]
```diff
@@ -1,13 +1,58 @@
...
+      .subscribe(() => this.enforceAttentionLevelRules());
+  }
+
+  ngOnDestroy() {
+    this.ngUnsubscribe$.next();
+    this.ngUnsubscribe$.complete();
+  }
+
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/empty-state/empty-state.component.ts:42]: No need to unsubscribe as this is handled by Angular when the component gets destroyed
```suggestion
```

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.ts [hunk 8]
```diff
@@ -1,13 +1,58 @@
...
+    let encounteredLvl1Button = false;
+
+    this.slottedButtons.forEach((button) => {
+      if (button.isAttentionLevel1 && !encounteredLvl1Button) {
+        encounteredLvl1Button = true;
+      } else if (!button.isAttentionLevel3) {
+        button.attentionLevel = '3';
+      }
+    });
+  }
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/empty-state/empty-state.component.ts:57]: The [issue/tech refinement](https://github.com/kirbydesign/designsystem/issues/1222#issuecomment-779841964) might not have been described well enough, but it's only supposed to be the 1st button in the list which is allowed attentionLevel1 (i.e. the wording "first of type" hinting at using `:first-of-type` pseudo selector)

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.ts [hunk 9]
```diff
@@ -1,13 +1,58 @@
...
+  ngOnDestroy() {
+    this.ngUnsubscribe$.next();
+    this.ngUnsubscribe$.complete();
+  }
+
+  /** Enforces that all buttons will have their attention
+   * level set to 3 except the first encountered button
+   * with level 1
+   */
+  private enforceAttentionLevelRules() {
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/empty-state/empty-state.component.ts:47]: Nice naming 👌

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.ts [hunk 10]
```diff
@@ -1,13 +1,58 @@
...
+      .pipe(takeUntil(this.ngUnsubscribe$), delay(0))
+      .subscribe(() => this.enforceAttentionLevelRules());
+  }
+
+  ngOnDestroy() {
+    this.ngUnsubscribe$.next();
+    this.ngUnsubscribe$.complete();
+  }
+
+  /** Enforces that all buttons will have their attention
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/empty-state/empty-state.component.ts:43]: Consider wording:
```suggestion
  /** Enforces that all slotted buttons will have their attention
```

**Review by jakobe — CHANGES_REQUESTED** [review]
The issue/tech refinement might not have been described well enough, but it's only supposed to be the 1st button in the list which is allowed attentionLevel1 (i.e. the wording "first of type" hinting at using :first-of-type pseudo selector)
Let's talk about tests :D

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.ts (hunk 9 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/empty-state/empty-state.component.ts:47]: Thank you 🙆‍♂️

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.ts (hunk 10 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/empty-state/empty-state.component.ts]: Should be solved by bd0ef64

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.ts (hunk 8 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/empty-state/empty-state.component.ts]: Should be resolved by 3f5f00b

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.ts (hunk 7 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/empty-state/empty-state.component.ts]: Should be resolved by 867ba3e

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.ts (hunk 6 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/empty-state/empty-state.component.ts]: Should be resolved by 867ba3e

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts (hunk 5 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts]: Should be resolved by 6703cee

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts (hunk 4 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts]: Should be resolved by 6703cee

File: apps/cookbook/src/app/examples/empty-state-example/empty-state-example.component.html (hunk 3 above)
Comment (MadsBuchmann) [reply apps/cookbook/src/app/examples/empty-state-example/empty-state-example.component.html]: Should be resolved by bd0ef64

Comment (MadsBuchmann) [reply apps/cookbook/src/app/examples/empty-state-example/empty-state-example.component.html]: Should be resolved by bd0ef64

**Comment by MadsBuchmann** [comment]
I've implemented the review suggestions. 

As I am testing how buttons behave when inserted into empty-state, i have moved the tests regarding the enforcement of attention level rules to `empty-state.component.integration.spec.ts`. 

I have reverted `empty-state.component.spec.ts` as these tests now are in `empty-state.component.integration.spec.ts`.

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts [hunk 11]
```diff
@@ -1,8 +1,9 @@
+import { createHostFactory, SpectatorHost } from '@ngneat/spectator';
 import { MockComponent } from 'ng-mocks';
-import { SpectatorHost, createHostFactory } from '@ngneat/spectator';
 
-import { DesignTokenHelper } from './../../helpers/design-token-helper';
 import { IconComponent } from '../icon/icon.component';
+
+import { DesignTokenHelper } from './../../helpers/design-token-helper';
```
Comment (MadsBuchmann) [inline libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts:6]: I've attempted to revert this file - but whenever I commit the imports are re-ordered automatically.

Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/empty-state/empty-state.component.spec.ts]: It managed to make it work!

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.integration.spec.ts [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    spectator = createHost(`
+          <kirby-empty-state
+              iconName="help"
+              title="No items"
+              subtitle="You don't have any items. Call support to add some items to your account."
+          >
+              <button kirby-button attentionLevel='4'>Call support</button>
+              <button kirby-button attentionLevel='3'>Mail support</button>
+              <button kirby-button attentionLevel='2'>Get directions</button>
+              <button kirby-button attentionLevel='1'>Cancel</button>
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/empty-state/empty-state.component.integration.spec.ts:28]: We generally use `"` for property assignment:
```suggestion
              <button kirby-button attentionLevel="4">Call support</button>
              <button kirby-button attentionLevel="3">Mail support</button>
              <button kirby-button attentionLevel="2">Get directions</button>
              <button kirby-button attentionLevel="1">Cancel</button>
```

Comment (jakobe) [inline libs/designsystem/src/lib/components/empty-state/empty-state.component.integration.spec.ts:28]: Consider not having attentionlevel set in the default test and move this to a separate test...

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.ts [hunk 13]
```diff
@@ -1,13 +1,39 @@
...
   @Input() title: string;
   @Input() subtitle: string;
+
+  @ContentChildren(ButtonComponent)
+  private slottedButtons: QueryList<ButtonComponent>;
+
+  ngAfterContentInit() {
+    this.enforceAttentionLevelRules();
+
+    this.slottedButtons.changes.pipe(delay(0)).subscribe(() => this.enforceAttentionLevelRules());
```
Comment (jakobe) [inline libs/designsystem/src/lib/components/empty-state/empty-state.component.ts:23]: Why the need for `delay`? Consider adding comment...

Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/empty-state/empty-state.component.ts:25]: Is this sufficient 😄?

Comment (jakobe) [reply libs/designsystem/src/lib/components/empty-state/empty-state.component.ts:25]: Yes much better :)
Consider stating why it's necessary as we do in RadioGroupComponent:
`// setTimeout prevents ExpressionChangedAfterItHasBeenCheckedError when updating the DOM in QueryList.changes:`
You could also consider wrapping the call to `this.enforceAttentionLevelRules` in a timeout instead, as this is the conventional way to postpone a state change to next CD...
See: https://github.com/kirbydesign/designsystem/blob/master/libs/designsystem/src/lib/components/radio/radio-group/radio-group.component.ts#L201

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.integration.spec.ts (hunk 12 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/empty-state/empty-state.component.integration.spec.ts]: I have moved them out to a separate test case `describe('EmptyStateComponent with slotted buttons configured with attention levels', () => {...` and the case you mention now uses defaults 😄

File: libs/designsystem/src/lib/components/empty-state/empty-state.component.ts (hunk 13 above)
Comment (MadsBuchmann) [reply libs/designsystem/src/lib/components/empty-state/empty-state.component.ts]: I've updated the comment to be: 
`/* setTimeout prevents ExpressionChangedAfterItHasBeenCheckedError when changing attention 
     levels of slotted buttons in this.enforceAttentionLevelRules */`

and removed the use of delay 😀

**Review by jakobe — APPROVED** [review]

**Pull request merged**